File: src/task_agent/store.py
from __future__ import annotations

import json
import os
import threading
import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from src.task_agent.definitions import DEFINITION_BY_TYPE, build_task_label

MAX_RECENT_TASKS = 20
_lock = threading.Lock()


def _repo_root() -> Path:
    return Path(__file__).resolve().parents[2]


def tasks_dir() -> Path:
    override = os.environ.get("SJKX_TASK_DIR")
    if override:
        return Path(override)
    return _repo_root() / "logs" / "ai-tasks"


def state_path() -> Path:
    return tasks_dir() / "state.json"


def _now_iso() -> str:
    return datetime.now(timezone.utc).astimezone().isoformat()


def _empty_state() -> dict[str, Any]:
    return {"tasks": [], "runningTaskId": None}


def _read_state_unlocked() -> dict[str, Any]:
    path = state_path()
    if not path.exists():
        return _empty_state()
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return _empty_state()
    if not isinstance(data.get("tasks"), list):
        return _empty_state()
    return {"tasks": data["tasks"], "runningTaskId": data.get("runningTaskId")}


def _write_state_unlocked(state: dict[str, Any]) -> None:
    directory = tasks_dir()
    directory.mkdir(parents=True, exist_ok=True)
    path = state_path()
    temp_path = path.with_suffix(".tmp")
    temp_path.write_text(json.dumps(state, ensure_ascii=False, indent=2), encoding="utf-8")
    temp_path.replace(path)


def create_task(task_type: str, params: dict[str, Any] | None = None) -> dict[str, Any]:
    params = params or {}
    definition = DEFINITION_BY_TYPE.get(task_type)
    if not definition:
        raise ValueError(f"未知任务类型: {task_type}")
    if definition.get("requiresTable") and not params.get("table"):
        raise ValueError("该任务需要指定竞品表（F / R / V）")

    task = {
        "id": str(uuid.uuid4()),
        "type": task_type,
        "label": build_task_label(task_type, params),
        "description": definition["description"],
        "params": params,
        "status": "pending",
        "createdAt": _now_iso(),
        "startedAt": None,
        "finishedAt": None,
        "logFile": None,
        "exitCode": None,
        "error": None,
    }

    with _lock:
        state = _read_state_unlocked()
        state["tasks"].insert(0, task)
        _write_state_unlocked(state)
    return task


def get_task(task_id: str) -> dict[str, Any] | None:
    with _lock:
        state = _read_state_unlocked()
        for task in state["tasks"]:
            if task["id"] == task_id:
                return task
    return None


def list_snapshot() -> dict[str, Any]:
    with _lock:
        state = _read_state_unlocked()
        tasks = state["tasks"]
        pending = [task for task in tasks if task.get("status") == "pending"]
        running = next(
            (
                task
                for task in tasks
                if task.get("id") == state.get("runningTaskId") and task.get("status") == "running"
            ),
            None,
        )
        if running is None:
            running = next((task for task in tasks if task.get("status") == "running"), None)
        recent = [
            task
            for task in tasks
            if task.get("status") in {"completed", "failed", "cancelled"}
        ][:MAX_RECENT_TASKS]
        return {"pending": pending, "running": running, "recent": recent}


def get_running_task_id() -> str | None:
    with _lock:
        return _read_state_unlocked().get("runningTaskId")


def set_running_task_id(task_id: str | None) -> None:
    with _lock:
        state = _read_state_unlocked()
        state["runningTaskId"] = task_id
        _write_state_unlocked(state)


def get_next_pending() -> dict[str, Any] | None:
    with _lock:
        state = _read_state_unlocked()
        return next((task for task in state["tasks"] if task.get("status") == "pending"), None)


def mark_running(task_id: str, log_file: str) -> dict[str, Any] | None:
    with _lock:
        state = _read_state_unlocked()
        for index, task in enumerate(state["tasks"]):
            if task["id"] != task_id:
                continue
            updated = {
                **task,
                "status": "running",
                "startedAt": _now_iso(),
                "logFile": log_file,
                "exitCode": None,
                "error": None,
            }
            state["tasks"][index] = updated
            state["runningTaskId"] = task_id
            _write_state_unlocked(state)
            return updated
    return None


def finalize_task(
    task_id: str,
    status: str,
    exit_code: int | None,
    error: str | None,
) -> dict[str, Any] | None:
    with _lock:
        state = _read_state_unlocked()
        if state.get("runningTaskId") == task_id:
            state["runningTaskId"] = None
        for index, task in enumerate(state["tasks"]):
            if task["id"] != task_id:
                continue
            updated = {
                **task,
                "status": status,
                "finishedAt": _now_iso(),
                "exitCode": exit_code,
                "error": error,
            }
            state["tasks"][index] = updated
            _write_state_unlocked(state)
            return updated
    return None


def cancel_pending(task_id: str) -> dict[str, Any] | None:
    with _lock:
        state = _read_state_unlocked()
        for index, task in enumerate(state["tasks"]):
            if task["id"] != task_id or task.get("status") != "pending":
                continue
            updated = {**task, "status": "cancelled", "finishedAt": _now_iso()}
            state["tasks"][index] = updated
            _write_state_unlocked(state)
            return updated
    return None

File: web/src/lib/ai-tasks/store.ts
import { existsSync, mkdirSync, readFileSync, renameSync, writeFileSync } from "node:fs";
import { randomUUID } from "node:crypto";

import { buildTaskLabel, getTaskDefinition } from "@/lib/ai-tasks/definitions";
import { getAiTasksDir, getAiTasksStatePath, isServerlessDeployment } from "@/lib/ai-tasks/paths";
import type { AiTask, AiTaskParams, AiTaskStatus, AiTaskType } from "@/lib/ai-tasks/types";

type TaskStoreState = {
  tasks: AiTask[];
  runningTaskId: string | null;
};

const MAX_RECENT_TASKS = 20;

function emptySnapshot() {
  return { pending: [] as AiTask[], running: null as AiTask | null, recent: [] as AiTask[] };
}

function emptyState(): TaskStoreState {
  return { tasks: [], runningTaskId: null };
}

function ensureStoreDir() {
  if (isServerlessDeployment()) return;
  mkdirSync(getAiTasksDir(), { recursive: true });
}

function readState(): TaskStoreState {
  if (isServerlessDeployment()) {
    return emptyState();
  }
  ensureStoreDir();
  const path = getAiTasksStatePath();
  if (!existsSync(path)) {
    return emptyState();
  }
  try {
    const raw = readFileSync(path, "utf8");
    const parsed = JSON.parse(raw) as TaskStoreState;
    if (!Array.isArray(parsed.tasks)) {
      return emptyState();
    }
    return {
      tasks: parsed.tasks,
      runningTaskId: parsed.runningTaskId ?? null,
    };
  } catch {
    return emptyState();
  }
}

function writeState(state: TaskStoreState) {
  if (isServerlessDeployment()) return;
  ensureStoreDir();
  const path = getAiTasksStatePath();
  const tempPath = `${path}.tmp`;
  writeFileSync(tempPath, JSON.stringify(state, null, 2), "utf8");
  renameSync(tempPath, path);
}

function mutateState(mutator: (state: TaskStoreState) => void): TaskStoreState {
  const state = readState();
  mutator(state);
  writeState(state);
  return state;
}

export function createTask(type: AiTaskType, params: AiTaskParams = {}): AiTask {
  if (isServerlessDeployment()) {
    throw new Error("Vercel 上无法创建任务");
  }

  const definition = getTaskDefinition(type);
  if (definition.requiresTable && !params.table) {
    throw new Error("该任务需要指定竞品表（F / R / V）");
  }

  const task: AiTask = {
    id: randomUUID(),
    type,
    label: buildTaskLabel(type, params),
    description: definition.description,
    params,
    status: "pending",
    createdAt: new Date().toISOString(),
    startedAt: null,
    finishedAt: null,
    logFile: null,
    exitCode: null,
    error: null,
  };

  mutateState((state) => {
    state.tasks.unshift(task);
  });

  return task;
}

export function getTaskById(taskId: string): AiTask | null {
  const state = readState();
  return state.tasks.find((task) => task.id === taskId) ?? null;
}

export function updateTask(taskId: string, patch: Partial<AiTask>): AiTask | null {
  let updated: AiTask | null = null;
  mutateState((state) => {
    const index = state.tasks.findIndex((task) => task.id === taskId);
    if (index === -1) return;
    updated = { ...state.tasks[index], ...patch };
    state.tasks[index] = updated;
  });
  return updated;
}

export function setRunningTaskId(taskId: string | null) {
  mutateState((state) => {
    state.runningTaskId = taskId;
  });
}

export function getRunningTaskId(): string | null {
  return readState().runningTaskId;
}

export function cancelPendingTask(taskId: string): AiTask | null {
  let cancelled: AiTask | null = null;
  mutateState((state) => {
    const index = state.tasks.findIndex((task) => task.id === taskId);
    if (index === -1) return;
    const task = state.tasks[index];
    if (task.status !== "pending") return;
    cancelled = {
      ...task,
      status: "cancelled",
      finishedAt: new Date().toISOString(),
    };
    state.tasks[index] = cancelled;
  });
  return cancelled;
}

export function finalizeTask(
  taskId: string,
  status: Extract<AiTaskStatus, "completed" | "failed">,
  exitCode: number | null,
  error: string | null,
): AiTask | null {
  let finalized: AiTask | null = null;
  mutateState((state) => {
    if (state.runningTaskId === taskId) {
      state.runningTaskId = null;
    }
    const index = state.tasks.findIndex((task) => task.id === taskId);
    if (index === -1) return;
    finalized = {
      ...state.tasks[index],
      status,
      finishedAt: new Date().toISOString(),
      exitCode,
      error,
    };
    state.tasks[index] = finalized;
  });
  return finalized;
}

export function listTasksSnapshot() {
  if (isServerlessDeployment()) {
    return emptySnapshot();
  }

  const state = readState();
  const pending = state.tasks.filter((task) => task.status === "pending");
  const running =
    state.tasks.find((task) => task.id === state.runningTaskId && task.status === "running") ??
    state.tasks.find((task) => task.status === "running") ??
    null;
  const recent = state.tasks
    .filter((task) => ["completed", "failed", "cancelled"].includes(task.status))
    .slice(0, MAX_RECENT_TASKS);

  return { pending, running, recent };
}

export function getNextPendingTask(): AiTask | null {
  const state = readState();
  return state.tasks.find((task) => task.status === "pending") ?? null;
}

export function markTaskRunning(taskId: string, logFile: string): AiTask | null {
  let running: AiTask | null = null;
  mutateState((state) => {
    const index = state.tasks.findIndex((task) => task.id === taskId);
    if (index === -1) return;
    running = {
      ...state.tasks[index],
      status: "running",
      startedAt: new Date().toISOString(),
      logFile,
      exitCode: null,
      error: null,
    };
    state.tasks[index] = running;
    state.runningTaskId = taskId;
  });
  return running;
}
